### PR TITLE
perf-tuning: add docs for subquery optimizations

### PR DESCRIPTION
**Pull request by winoros**
### What is changed, added or deleted? (Required)

Add docs for subquery related optimizations.

### Which TiDB version(s) do your changes apply to? (Required)

- [x] master (the latest development version)
- [x] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### CODE REVIEW

File: subquery-optimization.md [hunk 1]
```diff
@@ -3,4 +3,75 @@ title: 子查询相关的优化
...
+- `NOT EXISTS (SELECT ... FROM ...)`
+- `IN (SELECT ... FROM ..)`
+- `EXISTS (SELECT ... FROM ...)`
+- `... >/>=/</<=/=/!= (SELECT ... FROM ...)`
+
+有时，会遇到子查询中包含了非子查询中的列，如 `select * from t where t.a in (select * from t2 where t.b=t2.b)` 中子查询中的 `t.b` 便不是子查询中的列而是从子查询外面引入的列。这种子查询通常会被成为`关联子查询`，外部引入的列会被成为`关联列`，关联子查询相关的优化可以在[关联子查询去关联](/correlated-subquery-optimization.md)中查看。本节中我们主要关注不涉及关联列的子查询。
+
+子查询默认会以[理解 TiDB 执行计划](/query-execution-plan.md)中提到的 `semi join` 作为默认的执行方式，同时对于一些特殊的子查询，TiDB 会做一些逻辑上的替换使得查询可以获得更好的执行性能
+
+## `... < ALL (SELECT ... FROM ...)` 或者 ` ... > ANY (SELECT ... FROM ...)`
```
Comment (TomShawn) [inline subquery-optimization.md:20]:
```suggestion
## `... < ALL (SELECT ... FROM ...)` 或者 `... > ANY (SELECT ... FROM ...)`
```

File: subquery-optimization.md [hunk 2]
```diff
@@ -3,4 +3,75 @@ title: 子查询相关的优化
...
+| │ └─IndexReader_28           | 9990.00 | root      |                        | index:IndexFullScan_27                                                     |
+| │   └─IndexFullScan_27       | 9990.00 | cop[tikv] | table:t2, index:idx(a) | keep order:false, stats:pseudo                                             |
+| └─TableReader_11(Probe)      | 1.00    | root      |                        | data:TableRangeScan_10                                                     |
+|   └─TableRangeScan_10        | 1.00    | cop[tikv] | table:t1               | range: decided by [test.t2.a], keep order:false, stats:pseudo              |
++------------------------------+---------+-----------+------------------------+----------------------------------------------------------------------------+
+```
+
+这个改写会在 `IN` 子查询相对较小，而外部查询相对较大时产生更好的执行性能。因为不经过改写的情况下，我们无法使用以 t2 为驱动表的 `index join`。同时这里的弊端便是当改写删成的聚合无法被自动消去且 `t2` 表比较大时，反而会影响查询的性能。目前 TiDB 中使用 [tidb\_opt\_insubq\_to\_join\_and\_agg](/tidb-specific-system-variables.md#tidb_opt_insubq_to_join_and_agg) 变量来控制这个优化的打开与否。当遇到不合适这个优化的情况可以手动关闭。
+
+
```
Comment (TomShawn) [inline subquery-optimization.md:60]:
```suggestion
```

File: correlated-subquery-optimization.md [hunk 3]
```diff
@@ -4,3 +4,58 @@ category: performance
 ---
 
 # 关联子查询去关联
+
+在[子查询相关的优化](/subquery-optimization.md)中我们介绍了当没有关联列时，TiDB是如何处理子查询的。由于关联子查询解除关联依赖是一个比较复杂的工作，我们这里介绍一些简单的场景以及这个优化规则的适用范围。
```
Comment (ran-huang) [inline correlated-subquery-optimization.md:8]: Please avoid using words like “我们”.
Instead, you can say “本文档介绍……”.

File: correlated-subquery-optimization.md [hunk 4]
```diff
@@ -4,3 +4,58 @@ category: performance
...
+
```
Comment (ran-huang) [inline correlated-subquery-optimization.md:9]: Please add some level-2 headings to better strucuture this document.
Some examples are: `## 实例`, `## 原理`, `## 限制`.

File: correlated-subquery-optimization.md [hunk 5]
```diff
@@ -4,3 +4,58 @@ category: performance
...
+以 `select * from t1 where t1.a < (select sum(t2.a) from t2 where t2.b = t1.b)` 为例，这里子查询 `t1.a < (select sum(t2.a) from t2 where t2.b = t1.b)` 中涉及了关联列上的条件 `t2.b=t1.b`，不过恰好由于这是一个等值条件，因此我们可以将其等价的改写为 `select t1.* from t1, (select b, sum(a) sum_a from t2 group by b) t2 where t1.b = t2.b and t1.a < t2.sum_a;`这样一个关联子查询就被重新改写为 `JOIN` 的形式。
+
+TiDB 之所以要进行这样的改写，是因为关联子查询每次子查询执行时都是要和它的外部查询结果绑定的。在上面的例子中如果 `t1.a` 有一千万个值，那这个子查询就要被重复执行一千万次，因为 `t2.b=t1.b` 这个条件会随着 `t1.a` 值的不同而发生变化。当我们通过一些手段将关联依赖解除后，这个子查询就只需要被执行一次了。
+
+同时这种改写的弊端在于在关联没有被解除时，我们是可以使用关联列上的索引的，也就是说虽然这个子查询可能被重复执行多次，但是每次都可以使用索引过滤数据。而解除关联的变换上，通常是会导致关联列的位置发生改变而导致虽然子查询只被执行了一次，但是单次执行的时间会比没有解除关联时的单次执行时间长。因此在外部的值比较少的情况下，不解除关联依赖反而可能对执行性能更优帮助。我们可以通过[优化规则及表达式下推的黑名单](/blacklist-control-plan.md)中关闭`子查询去关联`优化规则的方式来关闭这个优化。
+
+```
+mysql> create table t1(a int, b int);
```
Comment (ran-huang) [inline correlated-subquery-optimization.md:17]: Please separate commands from results.

File: subquery-optimization.md [hunk 6]
```diff
@@ -3,4 +3,74 @@ title: 子查询相关的优化
 category: performance
 ---
 
-# 子查询相关的优化
\ No newline at end of file
+# 子查询相关的优化
+
```
Comment (ran-huang) [inline subquery-optimization.md:7]: Add one or two sentences to introduce what this document is about.
Then start a level-2 heading to illustrate the details.

File: subquery-optimization.md [hunk 7]
```diff
@@ -3,4 +3,74 @@ title: 子查询相关的优化
...
+## `... < ALL (SELECT ... FROM ...)` 或者 `... > ANY (SELECT ... FROM ...)`
+
+对于这种情况我们可以将 `ALL` 或者 `ANY` 用 `MAX` 以及 `MIN` 来代替。不过由于在表为空时，`MAX(EXPR)` 以及 `MIN(EXPR)` 的结果会为 NULL，其表现形式和 `EXPR` 是有 `NULL` 值的结果一样。以及外部表达式结果为 `NULL` 时也会影响表达式的最终结果，因此这里完整的改写会是如下的形式：
+
+- `t.id < all(select s.id from s)` 会被改写为 `t.id < min(s.id) and if(sum(s.id is null) != 0, null, true)`。
+- `t.id < any (select s.id from s)` 会被改写为 `t.id < max(s.id) or if(sum(s.id is null) != 0, null, false)`。
+
+## `... != ANY (SELECT ... FROM ...)`
+
+对于这种情况，当子查询中不同值的各种只有一种的话，那只要和这个值对比就可以了，如果子查询中不同值的个数多余1个，那么必然会有不相等的情况出现。因此这样的子查询可以采取如下的改写手段。
```
Comment (ran-huang) [inline subquery-optimization.md:29]:
```suggestion
对于这种情况，当子查询中不同值的各种只有一种的话，那只要和这个值对比就即可。如果子查询中不同值的个数多于 1 个，那么必然会有不相等的情况出现。因此这样的子查询可以采取如下的改写手段：
```

**Comment by ran-huang** [comment]
@winoros Please address comments. Thanks~

**Comment by ran-huang** [comment]
@lzmhhh123 PTAL. Thanks!

File: correlated-subquery-optimization.md [hunk 8]
```diff
@@ -4,3 +4,69 @@ category: performance
...
+|     └─HashAgg_27                       | 1.00     | root      |                        | funcs:sum(Column#10)->Column#7                                               |
+|       └─IndexLookUp_28                 | 1.00     | root      |                        |                                                                              |
+|         ├─IndexRangeScan_25(Build)     | 10.00    | cop[tikv] | table:t2, index:idx(b) | range: decided by [eq(test.t2.b, test.t1.b)], keep order:false, stats:pseudo |
+|         └─HashAgg_17(Probe)            | 1.00     | cop[tikv] |                        | funcs:sum(test.t2.a)->Column#10                                              |
+|           └─TableRowIDScan_26          | 10.00    | cop[tikv] | table:t2               | keep order:false, stats:pseudo                                               |
++----------------------------------------+----------+-----------+------------------------+------------------------------------------------------------------------------+
+```
+
+在执行了关闭规划规则的语句后，可以在 `IndexRangeScan_25(Build)` 的 `operator info` 中看到 `range: decided by [eq(test.t2.b, test.t1.b)]`，便是关联依赖未被解除时，可以使用关联条件进行索引范围查询的展示。
```
Comment (ran-huang) [inline correlated-subquery-optimization.md:72]: `便是关联依赖未被解除时`：`便是` -> `表示`？

File: subquery-optimization.md [hunk 9]
```diff
@@ -3,4 +3,74 @@ title: 子查询相关的优化
...
+
+通常我们遇到的会是如下情况的子查询：
+
+- `NOT IN (SELECT ... FROM ...)`
+- `NOT EXISTS (SELECT ... FROM ...)`
+- `IN (SELECT ... FROM ..)`
+- `EXISTS (SELECT ... FROM ...)`
+- `... >/>=/</<=/=/!= (SELECT ... FROM ...)`
+
+有时，会遇到子查询中包含了非子查询中的列，如 `select * from t where t.a in (select * from t2 where t.b=t2.b)` 中，子查询中的 `t.b` 便不是子查询中的列，而是从子查询外面引入的列。这种子查询通常会被成为`关联子查询`，外部引入的列会被成为`关联列`，关联子查询相关的优化可以在[关联子查询去关联](/correlated-subquery-optimization.md)中查看。本节中我们主要关注不涉及关联列的子查询。
```
Comment (ran-huang) [inline subquery-optimization.md:16]:
```suggestion
有时，会遇到子查询中包含了非子查询中的列，如 `select * from t where t.a in (select * from t2 where t.b=t2.b)` 中，子查询中的 `t.b` 便不是子查询中的列，而是从子查询外面引入的列。这种子查询通常会被称为`关联子查询`，外部引入的列会被称为`关联列`，关联子查询相关的优化可以在[关联子查询去关联](/correlated-subquery-optimization.md)中查看。本节中主要关注不涉及关联列的子查询。
```

File: subquery-optimization.md [hunk 10]
```diff
@@ -3,4 +3,74 @@ title: 子查询相关的优化
...
+
+## `... = ALL (SELECT ... FROM ...)`
+
+对于这种情况，当子查询中不同值的个数多于一种的话，那么这个表达式的结果必然为假。因此这样的子查询在 TiDB 中会改写为如下的形式：
+
+- `select * from t where t.id = all (select s.id from s)` 会被改写为 `select t.* from t, (select s.id, count(distinct s.id) as cnt_distinct from s) where (t.id = s.id and cnt_distinct <= 1)`
+
+## `... IN (SELECT ... FROM ...)`
+
+对于这种情况，我们会将其改写为 `IN` 的子查询改写为 `SELECT ... FROM ... GROUP ...` 的形式然后将 `IN` 改写为普通的 `JOIN` 的形式。
```
Comment (ran-huang) [inline subquery-optimization.md:41]:
```suggestion
对于这种情况，会将其改写为 `IN` 的子查询改写为 `SELECT ... FROM ... GROUP ...` 的形式，然后将 `IN` 改写为普通的 `JOIN` 的形式。
```

File: subquery-optimization.md (hunk 6 above)
Comment (winoros) [reply subquery-optimization.md:7]: 这里的话，去后面的二级标题就是围绕着 10-14行提到的子查询展开的。这个还需要在改些啥吗

File: correlated-subquery-optimization.md (hunk 8 above)
Comment (winoros) [reply correlated-subquery-optimization.md:72]: 啊 确实是便是的意思，又稍微修改了一下

File: subquery-optimization.md (hunk 6 above)
Comment (ran-huang) [reply subquery-optimization.md:7]: 可以加一个总括性的介绍，比如 “本文档介绍……”

**Pull request merged**